File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/api/Cdn.java
package app.zxtune.fs.api;

import android.net.Uri;
import app.zxtune.BuildConfig;

public class Cdn {

  private static final Uri ROOT = Uri.parse(BuildConfig.CDN_ROOT);

  public static Uri asma(String path) {
    return getRoot().path("browse/asma/" + path).build();
  }

  public static Uri hvsc(String path) {
    return getRoot().path("browse/hvsc/" + path).build();
  }

  public static Uri joshw(String catalogue, String path) {
    return getRoot().path("browse/joshw/" + catalogue + "/" + path).build();
  }

  public static Uri amp(int id) {
    return getRoot().path("download/amp/ids/" + id).build();
  }

  public static Uri modland(String id) {
    // really url-encoded paths
    return getRoot().encodedPath("download/modland" + id).build();
  }

  public static Uri modarchive(int id) {
    return getRoot().path("download/modarchive/ids/" + id).build();
  }

  public static Uri aminet(String path) {
    return getRoot().path("download/aminet/mods/" + path).build();
  }

  private static Uri.Builder getRoot() {
    return ROOT.buildUpon();
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/api/Cdn.kt
package app.zxtune.fs.api

import android.net.Uri
import app.zxtune.BuildConfig

private val ROOT = Uri.parse(BuildConfig.CDN_ROOT)

object Cdn {
    fun asma(path: String): Uri = root.path("browse/asma/$path").build()

    fun hvsc(path: String): Uri = root.path("browse/hvsc/$path").build()

    fun joshw(catalogue: String, path: String): Uri =
        root.path("browse/joshw/$catalogue/$path").build()

    fun amp(id: Int): Uri = root.path("download/amp/ids/$id").build()

    // really url-encoded paths
    fun modland(id: String): Uri = root.encodedPath("download/modland$id").build()

    fun modarchive(id: Int): Uri = root.path("download/modarchive/ids/$id").build()

    fun aminet(path: String): Uri = root.path("download/aminet/mods/$path").build()

    fun vgmrips(path: String): Uri = root.path("download/vgmrips/$path").build()

    fun scene(path: String): Uri = root.path("browse/scene/music/$path").build()

    private val root: Uri.Builder
        get() = ROOT.buildUpon()
}
